arrow-core-data/src/test/kotlin/arrow/core/EitherTest.kt

```diff
@@ -203,5 +210,83 @@ class EitherTest : UnitSpec() {
           Right(a).handleErrorWith { Right(b) } == Right(a)
       }
     }
+
+    "suspended Either.fx can bind immediate values" {
+      Gen.either(Gen.string(), Gen.int())
+        .random()
+        .take(1001)
+        .forEach { either ->
+          Either.fx<String, Int> {
+            val res = !either
+            res
+          } shouldBe either
+        }
+    }
+
+    "suspended Either.fx can bind suspended values" {
+      Gen.either(Gen.string(), Gen.int())
+        .random()
+        .take(1001)
+        .forEach { either ->
+          Either.fx<String, Int> {
+            val res = !(suspend {
+              sleep(1)
+              either
+            }).invoke()
+
+            res
+          } shouldBe either
+        }
+    }
+
+    "suspended Either.fx can safely handle immediate exceptions" {
+      Gen.bind(Gen.int(), Gen.throwable(), ::Pair)
+        .random()
+        .take(1001)
+        .forEach { (i, exception) ->
+          shouldThrow<Throwable> {
+            Either.fx<String, Int> {
+              val res = !Either.Right(i)
+              throw exception
+              res
+            } shouldBe Either.Right(i)
+          } shouldBe exception
+        }
+    }
+
+    "suspended Either.fx can bind suspended exceptions" {
+      Gen.bind(Gen.int(), Gen.throwable(), ::Pair)
+        .random()
+        .take(1001)
+        .forEach { (i, exception) ->
+          shouldThrow<Throwable> {
+            Either.fx<String, Int> {
+              val res = !Either.Right(i)
+              sleep(1)
+              throw exception
```

Comment:
I don't see the difference with the test above except for the sleep, which I'm unsure how would change the result. Shouldn't we test the Left binding instead? 🤔

Reply:
You need to test function that immediately return, and functions that `COROUTINE_SUSPENDED` before they return.

`sleep(1)` guarantees that `COROUTINE_SUSPENDED` is returned [here](https://github.com/arrow-kt/arrow-core/pull/112/files#diff-ffdbb5ceda6c9cb4fc2a0f86a81eaddfR1111).

We test both `Left` and `Right` since we generate `Either`, immediately bind and check that the operation is the same as `identity`.